proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,230 @@
+
+* Quasiquotes induces a dependency on Template Haskell, which a lot of people avoid out of principle. Most QuasiQuoters also add a dependency on ``haskell-src-exts`` to parse arbitrary Haskell expressions, which could technically be avoided by using something like ``ghc-meta`` (`repo <https://github.com/noughtmare/ghc-meta>`_, `GHC issue <https://gitlab.haskell.org/ghc/ghc/-/issues/20862>`_), but this isn't in wide use yet.
+
+If Haskell had native string interpolation, it would have the benefit and safety of the current third-party quasiquotes without the need for Template Haskell, and be able to take advantage of features like `multiline strings <https://github.com/ghc-proposals/ghc-proposals/pull/569>`_.
+::
+
+  error s"Expected: ${x + y}, got: ${result}"
+
+  let textExample = s"${name1} (age: ${age1}) encountered ${name2} (age: ${age2})"
+
+Proposed Change Specification
+-----------------------------
+
+When a new ``StringInterpolation`` extension is enabled, the following syntax:
+::
+
+  s"foo ${f a b} bar ${g x} baz ${name}"
+
+will be syntax sugar for:
+::
+
+  interpolateRaw "foo "
+  . doInterpolate (f a b)
+  . interpolateRaw " bar "
+  . doInterpolate (g x)
+  . interpolateRaw " baz "
+  . doInterpolate name
+  $ interpolateEmpty
+
+where:
+::
+
+  class Interpolate s where
+    interpolateRaw :: String -> s -> s
+    default interpolateRaw :: (IsString s, Semigroup s) => String -> s -> s
+    interpolateRaw s = (fromString s <>)
+
+    interpolateEmpty :: s
+    default interpolateEmpty :: IsString s => s
+    interpolateEmpty = fromString ""
+
+  class Interpolate s => InterpolateValueRunner s a where
+    doInterpolate :: a -> s -> s
+
+  class Interpolate s => InterpolateValue flag s a where
+    {-# MINIMAL interpolate | interpolatePrec #-}
+
+    interpolate :: Proxy flag -> a -> s
+    interpolate proxy a = interpolatePrec proxy 0 a interpolateEmpty
+
+    interpolatePrec :: Proxy flag -> Int -> a -> s -> s
+    default interpolatePrec :: Semigroup s => Proxy flag -> Int -> a -> s -> s
+    interpolatePrec proxy _ a s = interpolate proxy a <> s
+
+  data InterpolateDefault
+  data InterpolateOverride
+
+  class InterpolateValueFlag s a flag | s a -> flag
+  instance {-# OVERLAPPABLE #-} (flag ~ InterpolateDefault) => InterpolateValueFlag s a flag
+
+  instance
+    ( Interpolate s
+    , InterpolateValueFlag s a flag
+    , InterpolateValue flag s a
+    ) => InterpolateValueRunner s a where
+    doInterpolate = interpolatePrec (Proxy :: Proxy flag) 0
+
+We will always require interpolation to be of the form ``${foo}``, unlike other languages where ``$foo`` is allowed for single-variable interpolation. This simplifies the syntax and avoids the need for escaping bare ``$``. Interpolation may be escaped by escaping the dollar sign; e.g. ``"\${foo}"`` is equivalent to ``['$', '{', 'f', 'o', 'o', '}']``.
```

Comment:
Could you share the design decision behind using explicit interpolation delimiter `$`?  As aforementioned, python, rust and [PyF](https://hackage.haskell.org/package/PyF) only uses `{}` to mark a placeholder, and (albeit subjective) looks cleaner to me. Did adding explicit delimiter make implementation easier?

Reply:
It shouldnt have any effect on implementation. Personally I like it, as curly braces are fairly common, while the `${` string is much less likely to come up.

e.g. if you want to interpolate into a JSON string:

```
s"{ "name": ${name} }"
```
vs
```
s"\{ "name": {name} \}"
```

Reply:
Thanks for the quick answer. Indeed the benefits of not having to escape brackets would outweigh the inconveniences of having to type an extra `$`.

Reply:
Might also be worth noting that it looks syntactically similar to the way we splice in Template Haskell quotes. I think this is nice, in that both of them have the quality of "insert this computed Haskell value into the surrounding context".

Indeed, I wonder if we could make the syntax even closer and use `$()` instead of `${}`?

Reply:
I'd rather take the precendent of basically every other language, including e.g. bash and stay with curly braces

Reply:
FWIW you could generalise $() even further to $aexp e.g. $name, with $(exp) being an instance of the rule, then $ would just be escaped as \$.

Reply:
> I'd rather take the precendent of basically every other language, including e.g. bash and stay with curly braces

Sure, but the _closest_ precedent is the precedent of the language we are already in, i.e. Haskell :shrug: